### DISCUSSION TITLE
#12 - Testing (our patience) and other random stuff ❄🧪

### DISCUSSION THREAD

**Discussion by AlexMi-Ha** [CUMGroup/AMOGUS#103] [Blog]
# Welcome back fellow **AMOGUS** enjoyers 🔥
![You](https://media.tenor.com/9dtewBGJxlsAAAAd/dancing-robin.gif)
> ↳ You probably right now 

Its been a long week and a lot of work was done! This weeks topic is Testing! Yaaaaay!
<img src="https://user-images.githubusercontent.com/61502536/235673730-558d3bc0-fa02-41ee-a303-b02e830edbb0.png" alt="You now" width=200></img>
> ↳ You after hearing the word 'testing'

Even though testing can be kind of annoying it is an essential part of the Software Engineering process! It ensures delivering good, functional, working end results and helps developers to check if changes had some impact on the application.⚙

In our project we are using the [xUnit](https://github.com/xunit/xunit) testing framework to create our unit test cases. As we use a _Service bases Architecture_ we also have to mock all dependencies of the _system under test_ (i.e. other Services, Date providers...). For this we use the [Moq](https://github.com/moq/moq4) library which makes it easy to implement partial classes for each dependency interface.

An example Testcase would look something like this:
```csharp
[Fact]
public async Task CreateRolesAsync_WhenGivenRoles_CreatesThem() {
	// Arrange
	var createdIdentityRoles = new List<IdentityRole>();
	var roleManager = CreateRoleManagerMock();
	roleManager
		.Setup(x => x.RoleExistsAsync(It.IsAny<string>()))
		.ReturnsAsync(false);
	roleManager
		.Setup(x => x.CreateAsync(It.IsAny<IdentityRole>()))
		.Callback((IdentityRole ir) => createdIdentityRoles.Add(ir));
	
	var authService = new AuthService(roleManager.Object, CreateUserManagerMock().Object, CreateTokenFactoryMock().Object, CreateUserStatsRepositoryMock().Object, CreateValidatorMock().Object);

	// Act
	await authService.CreateRolesAsync<UserRoles>();

	// Assert
	Assert.Contains(createdIdentityRoles, x => "Admin".Equals(x.Name));
	Assert.Contains(createdIdentityRoles, x => "Moderator".Equals(x.Name));
	Assert.Contains(createdIdentityRoles, x => "User".Equals(x.Name));
}
```
As you can see we first Arrange (setup all dependencies...), then Act (execute the method) and finally Assert (test if the output is as expected).

Within Visual Studio we can then simply execute all tests in the current assembly and get an overview over the results
![image](https://user-images.githubusercontent.com/61502536/235677192-d1cc759e-236d-47d3-aa3b-e26832b4f6d8.png)

To get an better overview over the tests and compare with past results, we can use the exported Test Summary ✏

![image](https://user-images.githubusercontent.com/61502536/235679380-144cd7a7-840f-4513-8e62-3b13c935d04a.png)

Here we can also get an overview over our code coverage

![image](https://user-images.githubusercontent.com/61502536/235680250-ac30dbd9-43e9-4f38-99da-c3f9a604f160.png)

**You can find our detailed test plan [here](https://cumgroup.github.io/AMOGUS/Developer%20Docs/Testing/Testplan/)**

For the next Sprint we plan to write even more tests to achieve a better coverage. Furthermore we will include the tests in our CI process. But more on that in next weeks blog post ^^

So... Enough testing for today! 😌

We also worked on our core gameplay a lot this week!
**AMOGUS** is now capable of not only delivering pre defined questions, but we can now generate mental-exercises on the fly. This was done in two difficulty ranges - Randomized Mental Mode & Randomized Mental **Insane** Mode 💀
> The name keeps what it says - It is insanely difficult (but also gives a lot more experience points soooo 👀)

That's it with this weeks blog post!
> Stay sus!

Yours truly,
CUMGroup

**Comment by tonoizer** [comment]
Hey,

this looks great. Are you satisfied with the xUnit Test Framework, I have never used it but it goods quite simple to use and the text coverage summary report looks good too.
It looks like you have already covered quite some code with your tests, which is good to see. but also a lot of uncovered lines which should be covered in the future.

Keep up the  work

Team Vigad

**Reply by AlexMi-Ha** [reply]
Hi Team Vigad,
we are quite happy with the xUnit Framework. It's really easy to set up (2 clicks in Visual studio) and then just works. Writing tests and assertion is also really easy with a bunch of really helpful ExtensionMethods.
We are working on covering those last few lines (already up to 72% line coverage at this time 🥳) 

Yours truly,
CUMGroup

**Comment by Fullstack-DEVil** [comment]
Hey AMOGUS Team,

Your blog entry sounds very promising. The xUnit framework is also something new for me. Nice that you added a sample code from your test. Some development teams here can definitely benefit from it. Your test results look great. I also had a look at your test plan. It is very clear, organized and very well thought out.

Good luck with the further testing
 
Your Rule The Gym - Team

**Reply by AlexMi-Ha** [reply]
Hello Rule The Gym,
thank you 😊 
It is quite easy to get started in this framework in my opinion. It's always good to get inspiration and references from others ^^

Yours truly,
CUMGroup

**Comment by WeberWr** [comment]
Hey AMOGUS,

your blog always makes our day better. We think you did a good job integrating xUnit to your project. Your test plan looks good and understandable. You do a good job combine interesting facts with funny jokes. Good luck 🍀  with further testing. 
I think other teams can use this blog as example.

Regards
TD-Team

**Reply by AlexMi-Ha** [reply]
Hi TD-Team,
great to hear your feedback 😊
We always try to make our blogs enjoyable to read to keep things fresh! ❄🍃

We also wish you good luck with your testing.🤗

Yours truly,
CUMGroup

**Comment by vans1646** [comment]
Hi CUMGroup,
another week, another great blogpost from you :)

It's very good, that you know how important testing is for a software. Is the xUnit framework similar to JUnit, not for Java but for any other language? Your example test has a similar structure and I think it's very easy to use and functional.

I also think it's very good to integrate the tests into your CI so you don't have to run the tests manually and maybe during holidays miss an important bug. In our team we use GitHub Actions so considering that you also use GitHub, that could maybe be a good option. If you also use Jira for your Scrum workflow you can also connect these processes and have a nice overview on your Jira board.
Otherwise I can also recommend doing it via Jenkins, which has an easy setup and can be used in various scenarios.

Keep the good work up and glad that you also programmed more functionality of your software.

Cheers,
Capyclue

**Reply by AlexMi-Ha** [reply]
Hi Capyclue 😊
always nice to hear from you! ^^

With xUnit we meant xUnit.net in particular. It's the standard Testing Framework in C#, F# and VB. From the Idea it's similar to JUnit as both are based on the xUnit architecture.
But in my opinion xUnit.net is much better to write as a lot of features are solved with dedicated generic paramerters or lambdas/predicates instead of assert methods with lengthy names or strings.
It just seems a lot more polished.

We are also using Github Actions for CI, where every pull request gets tested. The result and code coverage gets collected in the Actions artifact.
![image](https://user-images.githubusercontent.com/61502536/236527078-2f57b633-ee50-46a7-b34d-40f954dc24b1.png)

How did you connect your tests to Jira to get a better overview than in the action logs? 🤔

Yours truly,
CUMGroup

**Reply by vans1646** [reply]
Very useful tool xUnit, and as I've seen in your latest blog post you also know how to use it properly, since you tested nearly everything :)
To connect Jira with GitHub you can use the Plugin 'GitHub for Jira' from the Atlassian Marketplace. If you need more info you can have a look [here](https://support.atlassian.com/jira-cloud-administration/docs/integrate-github-with-jira/)

Cheers :)

**Comment by Richone01** [comment]
Hi team AMOGUS,

Your testing Strategy looks great! I think it's very inzeresting what frameworks you use, as our project (DHBWorkout) has kind of the same Architecture (we're also service based) but use different ones (JUnit and Mockito). But it looks like yours is working as good and I really like to see that there are also other ways to implement it!

Good luck with your further testing!
Leon from DHBWorkout
